### ISSUE TITLE
Delaunay triangulation is invalid on data that Scipy handles correctly

### ISSUE THREAD

**Issue by mpetitjean** [matplotlib/matplotlib#16912]
### Bug report

**Bug summary**

With some of my data, the Delaunay triangulation of matplotlib creates an invalid triangulation. Running `scipy.spatial.Delaunay` with the same data yields no error, hinting that the data seems to be valid.

**Code for reproduction**

The following code creates a triangulation based on my data and attempts to refine it. The content of `data.csv` is [here](https://pastebin.com/b91vSJk4).

```python
import matplotlib.pyplot as plt
import matplotlib.tri as tri
from scipy.spatial import Delaunay
import numpy as np

# extract data
data = np.loadtxt("data.csv", delimiter=";")
x = data[:,0]
y = data[:,1]
t = data[:,2]

# compute triangulation with matlplotlib.tri
triang = tri.Triangulation(x,y)

# masking badly shaped triangles at the border of the triangular mesh.
# see https://matplotlib.org/3.1.1/gallery/images_contours_and_fields/tricontour_smooth_delaunay.html
min_circle_ratio = .01
mask = tri.TriAnalyzer(triang).get_flat_tri_mask(min_circle_ratio)
triang.set_mask(mask)

# refining the data
refiner = tri.UniformTriRefiner(triang)
triang_ref, t_ref = refiner.refine_field(t, subdiv=3)
```

**Actual outcome**

The triangulation is invalid. Console output is:

```
C:\Users\Mathieu\Anaconda3\envs\pyinst\lib\site-packages\matplotlib\tri\tritools.py:110: RuntimeWarning: invalid value encountered in sqrt
  4.0*np.sqrt(prod[~bool_flat]))
C:\Users\Mathieu\Anaconda3\envs\pyinst\lib\site-packages\matplotlib\tri\tritools.py:172: RuntimeWarning: invalid value encountered in less
  mask_bad_ratio = self.circle_ratios(rescale) < min_circle_ratio
Traceback (most recent call last):
  File "d:\MEGA\Haemers-tech\triang_minimal\example.py.py", line 27, in <module>
    triang_ref, t_ref = refiner.refine_field(t, subdiv=3)
  File "C:\Users\Mathieu\Anaconda3\envs\pyinst\lib\site-packages\matplotlib\tri\trirefine.py", line 166, in refine_field
    self._triangulation, z)
  File "C:\Users\Mathieu\Anaconda3\envs\pyinst\lib\site-packages\matplotlib\tri\triinterpolate.py", line 389, in __init__
    TriInterpolator.__init__(self, triangulation, z, trifinder)
  File "C:\Users\Mathieu\Anaconda3\envs\pyinst\lib\site-packages\matplotlib\tri\triinterpolate.py", line 46, in __init__
    self._trifinder = trifinder or self._triangulation.get_trifinder()
  File "C:\Users\Mathieu\Anaconda3\envs\pyinst\lib\site-packages\matplotlib\tri\triangulation.py", line 179, in get_trifinder
    self._trifinder = TrapezoidMapTriFinder(self)
  File "C:\Users\Mathieu\Anaconda3\envs\pyinst\lib\site-packages\matplotlib\tri\trifinder.py", line 41, in __init__
    self._initialize()
  File "C:\Users\Mathieu\Anaconda3\envs\pyinst\lib\site-packages\matplotlib\tri\trifinder.py", line 85, in _initialize
    self._cpp_trifinder.initialize()
RuntimeError: In initialize: Triangulation is invalid
```

**Expected outcome**

A valid triangulation. The following code, using the same data and `scipy.spatial.Delaunay` runs fine.

```python
import matplotlib.pyplot as plt
import matplotlib.tri as tri
from scipy.spatial import Delaunay
import numpy as np

# extract data
data = np.loadtxt("data.csv", delimiter=";")
x = data[:,0]
y = data[:,1]
t = data[:,2]

# compute triangulation with scipy
scipy_triang = Delaunay(np.vstack((x, y)).T)
triang = tri.Triangulation(x, y, triangles=scipy_triang.simplices)

# masking badly shaped triangles at the border of the triangular mesh.
# see https://matplotlib.org/3.1.1/gallery/images_contours_and_fields/tricontour_smooth_delaunay.html
min_circle_ratio = .01
mask = tri.TriAnalyzer(triang).get_flat_tri_mask(min_circle_ratio)
triang.set_mask(mask)

# refining the data
refiner = tri.UniformTriRefiner(triang)
triang_ref, t_ref = refiner.refine_field(t, subdiv=3)
```

**Matplotlib version**

  * Operating system: Windows 10 Pro, Python and libraries installed via conda
  * Matplotlib version: 3.1.3
  * Matplotlib backend (`print(matplotlib.get_backend())`): Qt5Agg
  * Python version: 3.6.10
  * Other libraries: scipy 1.4.1

**Comment by mpetitjean** [comment]
Please note that for the code to run, [this patch](https://github.com/matplotlib/matplotlib/pull/12807) needs to be applied to matplotlib.

**Comment by mpetitjean** [comment]
Tried after upgrading to Python 3.8.2 and matplotlib 3.2.1 in a new virtual environment, same behavior.

**Comment by bedumont** [comment]
Both variants work as expected for me

**Versions**
* python: 3.8.2
* matplotlib: 3.2.1
* scipy 1.4.1
* os: arch

**Comment by ianthomas23** [comment]
This was fixed by PR #17358 which is in mpl version 3.2.2.
